src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/JenkinsService.java

```diff
@@ -465,6 +507,72 @@ private void addFeedbackToResult(Result result, TestResultsDTO report) {
         }
     }
 
+    private List<BuildLogEntry> parsePipelineLogs(Element logHtml) throws IllegalArgumentException {
+        final var buildLog = new LinkedList<BuildLogEntry>();
+        if (logHtml.childNodes().stream().noneMatch(child -> child.attr("class").contains("pipeline"))) {
+            throw new IllegalArgumentException("Log is not pipeline log");
+        }
+        for (Element elem : logHtml.children()) {
+            // Only pipeline-node-ID elements contain actual log entries
+            if (elem.attributes().get("class").contains("pipeline-node")) {
+                // At least one child must have a timestamp class
+                if (elem.childNodes().stream().anyMatch(child -> child.attr("class").contains("timestamp"))) {
+                    Iterator<Node> nodeIterator = elem.childNodes().iterator();
+
+                    while (nodeIterator.hasNext()) {
+                        Node node = nodeIterator.next();
+                        final String log;
+                        if (node.attributes().get("class").contains("timestamp")) {
+                            final var timeAsString = ((TextNode) node.childNode(0).childNode(0)).getWholeText();
+                            final var time = ZonedDateTime.parse(timeAsString, DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ssX"));
```

Comment:
DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ssX") could move this out of the loops.

Reply:
Moved it into the class.